File: src/memory/memory_layers.py
"""
Memory Layer Handlers - Four-Layer Memory Architecture

Implements IMemoryLayer Protocol with 4 layer handlers:
- EphemeralLayer: Temporary memory (< 1 hour TTL)
- SessionLayer: Current task scope
- UserLayer: Long-term user preferences
- ProjectLayer: Novel-level scope

Each layer has different TTL, scope, and persistence characteristics.
"""

import logging
from abc import abstractmethod
from dataclasses import dataclass, field
from datetime import datetime, timedelta
from enum import Enum
from typing import Any, Dict, List, Optional, Protocol, runtime_checkable

logger = logging.getLogger("niko-memory-layers")


class LayerType(Enum):
    """Memory layer types with TTL characteristics."""
    EPHEMERAL = "ephemeral"   # < 1 hour
    SESSION = "session"       # Current task
    USER = "user"             # Long-term preferences
    PROJECT = "project"       # Novel-level scope


@dataclass
class LayerConfig:
    """Configuration for a memory layer."""
    layer_type: LayerType
    default_ttl_seconds: Optional[int] = None  # None = no expiration
    max_entries: int = 10000
    auto_expire: bool = True
    persist_on_close: bool = True


# Default configurations for each layer
LAYER_CONFIGS: Dict[LayerType, LayerConfig] = {
    LayerType.EPHEMERAL: LayerConfig(
        layer_type=LayerType.EPHEMERAL,
        default_ttl_seconds=3600,  # 1 hour
        max_entries=1000,
        auto_expire=True,
        persist_on_close=False
    ),
    LayerType.SESSION: LayerConfig(
        layer_type=LayerType.SESSION,
        default_ttl_seconds=86400,  # 24 hours
        max_entries=5000,
        auto_expire=True,
        persist_on_close=True
    ),
    LayerType.USER: LayerConfig(
        layer_type=LayerType.USER,
        default_ttl_seconds=None,  # No expiration
        max_entries=10000,
        auto_expire=False,
        persist_on_close=True
    ),
    LayerType.PROJECT: LayerConfig(
        layer_type=LayerType.PROJECT,
        default_ttl_seconds=None,  # No expiration
        max_entries=50000,
        auto_expire=False,
        persist_on_close=True
    ),
}


@dataclass
class LayerEntry:
    """Entry stored in a memory layer."""
    id: str
    content: str
    created_at: datetime = field(default_factory=datetime.now)
    expires_at: Optional[datetime] = None
    importance: float = 0.5
    access_count: int = 0
    last_accessed: Optional[datetime] = None
    metadata: Dict[str, Any] = field(default_factory=dict)

    def is_expired(self) -> bool:
        """Check if entry has expired."""
        if self.expires_at is None:
            return False
        return datetime.now() > self.expires_at

    def touch(self) -> None:
        """Update access tracking."""
        self.access_count += 1
        self.last_accessed = datetime.now()


@dataclass
class LayerStats:
    """Statistics for a memory layer."""
    layer_type: LayerType
    total_entries: int
    expired_entries: int
    total_access_count: int
    avg_importance: float
    oldest_entry: Optional[datetime]
    newest_entry: Optional[datetime]


@runtime_checkable
class IMemoryLayer(Protocol):
    """
    Protocol for memory layer implementations.

    Each layer handles storage, retrieval, and expiration
    of memories within its scope.
    """

    @property
    def layer_type(self) -> LayerType:
        """Get the layer type."""
        ...

    @property
    def config(self) -> LayerConfig:
        """Get layer configuration."""
        ...

    async def store(
        self,
        id: str,
        content: str,
        importance: float = 0.5,
        ttl_seconds: Optional[int] = None,
        metadata: Optional[Dict[str, Any]] = None
    ) -> LayerEntry:
        """
        Store content in this layer.

        Args:
            id: Unique identifier for the entry.
            content: Content to store.
            importance: Importance score (0.0-1.0).
            ttl_seconds: Time-to-live in seconds (overrides default).
            metadata: Additional metadata.

        Returns:
            Created LayerEntry.
        """
        ...

    async def retrieve(self, id: str) -> Optional[LayerEntry]:
        """
        Retrieve an entry by ID.

        Args:
            id: Entry identifier.

        Returns:
            LayerEntry or None if not found/expired.
        """
        ...

    async def expire(self, id: str) -> bool:
        """
        Manually expire an entry.

        Args:
            id: Entry identifier.

        Returns:
            True if expired, False if not found.
        """
        ...

    async def expire_all(self) -> int:
        """
        Expire all entries that have passed their TTL.

        Returns:
            Number of entries expired.
        """
        ...

    async def stats(self) -> LayerStats:
        """
        Get layer statistics.

        Returns:
            LayerStats with current state.
        """
        ...

    async def list_entries(
        self,
        limit: int = 100,
        include_expired: bool = False
    ) -> List[LayerEntry]:
        """
        List entries in this layer.

        Args:
            limit: Maximum entries to return.
            include_expired: Whether to include expired entries.

        Returns:
            List of LayerEntry objects.
        """
        ...

    async def clear(self) -> int:
        """
        Clear all entries from this layer.

        Returns:
            Number of entries cleared.
        """
        ...


class BaseMemoryLayer:
    """
    Base implementation for memory layers.

    Provides common functionality for all layer types.
    """

    def __init__(self, config: LayerConfig):
        self._config = config
        self._entries: Dict[str, LayerEntry] = {}
        logger.info(f"Initialized {config.layer_type.value} layer")

    @property
    def layer_type(self) -> LayerType:
        return self._config.layer_type

    @property
    def config(self) -> LayerConfig:
        return self._config

    async def store(
        self,
        id: str,
        content: str,
        importance: float = 0.5,
        ttl_seconds: Optional[int] = None,
        metadata: Optional[Dict[str, Any]] = None
    ) -> LayerEntry:
        # Calculate expiration
        expires_at = None
        ttl = ttl_seconds or self._config.default_ttl_seconds
        if ttl is not None:
            expires_at = datetime.now() + timedelta(seconds=ttl)

        entry = LayerEntry(
            id=id,
            content=content,
            importance=importance,
            expires_at=expires_at,
            metadata=metadata or {}
        )

        # Check capacity
        if len(self._entries) >= self._config.max_entries:
            await self._evict_oldest()

        self._entries[id] = entry
        logger.debug(f"Stored entry {id} in {self.layer_type.value} layer")
        return entry

    async def retrieve(self, id: str) -> Optional[LayerEntry]:
        entry = self._entries.get(id)
        if entry is None:
            return None

        if entry.is_expired():
            if self._config.auto_expire:
                del self._entries[id]
            return None

        entry.touch()
        return entry

    async def expire(self, id: str) -> bool:
        if id in self._entries:
            del self._entries[id]
            logger.debug(f"Expired entry {id} from {self.layer_type.value} layer")
            return True
        return False

    async def expire_all(self) -> int:
        expired_ids = [
            id for id, entry in self._entries.items()
            if entry.is_expired()
        ]
        for id in expired_ids:
            del self._entries[id]

        if expired_ids:
            logger.info(f"Expired {len(expired_ids)} entries from {self.layer_type.value} layer")
        return len(expired_ids)

    async def stats(self) -> LayerStats:
        entries = list(self._entries.values())
        expired_count = sum(1 for e in entries if e.is_expired())
        total_access = sum(e.access_count for e in entries)
        avg_importance = (
            sum(e.importance for e in entries) / len(entries)
            if entries else 0.0
        )

        created_times = [e.created_at for e in entries]
        oldest = min(created_times) if created_times else None
        newest = max(created_times) if created_times else None

        return LayerStats(
            layer_type=self.layer_type,
            total_entries=len(entries),
            expired_entries=expired_count,
            total_access_count=total_access,
            avg_importance=avg_importance,
            oldest_entry=oldest,
            newest_entry=newest
        )

    async def list_entries(
        self,
        limit: int = 100,
        include_expired: bool = False
    ) -> List[LayerEntry]:
        entries = list(self._entries.values())

        if not include_expired:
            entries = [e for e in entries if not e.is_expired()]

        # Sort by importance (descending), then by created_at (descending)
        entries.sort(key=lambda e: (e.importance, e.created_at), reverse=True)
        return entries[:limit]

    async def clear(self) -> int:
        count = len(self._entries)
        self._entries.clear()
        logger.info(f"Cleared {count} entries from {self.layer_type.value} layer")
        return count

    async def _evict_oldest(self) -> None:
        """Evict oldest entry to make room."""
        if not self._entries:
            return

        # Find oldest entry
        oldest_id = min(
            self._entries.keys(),
            key=lambda id: self._entries[id].created_at
        )
        del self._entries[oldest_id]
        logger.debug(f"Evicted oldest entry {oldest_id} from {self.layer_type.value} layer")


class EphemeralLayer(BaseMemoryLayer):
    """
    Ephemeral memory layer.

    Short-lived memories with < 1 hour TTL.
    Used for temporary context, scratch data, and intermediate results.
    """

    def __init__(self, config: Optional[LayerConfig] = None):
        super().__init__(config or LAYER_CONFIGS[LayerType.EPHEMERAL])

    async def store(
        self,
        id: str,
        content: str,
        importance: float = 0.5,
        ttl_seconds: Optional[int] = None,
        metadata: Optional[Dict[str, Any]] = None
    ) -> LayerEntry:
        # Ephemeral layer enforces max TTL of 1 hour
        max_ttl = 3600
        if ttl_seconds is not None:
            ttl_seconds = min(ttl_seconds, max_ttl)
        else:
            ttl_seconds = self._config.default_ttl_seconds or max_ttl

        return await super().store(id, content, importance, ttl_seconds, metadata)


class SessionLayer(BaseMemoryLayer):
    """
    Session memory layer.

    Task-scoped memories that persist for the current session.
    Used for conversation context, task state, and working memory.
    """

    def __init__(
        self,
        session_id: Optional[str] = None,
        config: Optional[LayerConfig] = None
    ):
        super().__init__(config or LAYER_CONFIGS[LayerType.SESSION])
        self.session_id = session_id

    async def store(
        self,
        id: str,
        content: str,
        importance: float = 0.5,
        ttl_seconds: Optional[int] = None,
        metadata: Optional[Dict[str, Any]] = None
    ) -> LayerEntry:
        # Add session_id to metadata
        metadata = metadata or {}
        if self.session_id:
            metadata["session_id"] = self.session_id

        return await super().store(id, content, importance, ttl_seconds, metadata)


class UserLayer(BaseMemoryLayer):
    """
    User memory layer.

    Long-term user preferences and learned behaviors.
    No automatic expiration, persists across sessions.
    """

    def __init__(
        self,
        user_id: Optional[str] = None,
        config: Optional[LayerConfig] = None
    ):
        super().__init__(config or LAYER_CONFIGS[LayerType.USER])
        self.user_id = user_id

    async def store(
        self,
        id: str,
        content: str,
        importance: float = 0.5,
        ttl_seconds: Optional[int] = None,
        metadata: Optional[Dict[str, Any]] = None
    ) -> LayerEntry:
        # User layer typically doesn't expire
        if ttl_seconds is None:
            ttl_seconds = self._config.default_ttl_seconds  # None

        metadata = metadata or {}
        if self.user_id:
            metadata["user_id"] = self.user_id

        return await super().store(id, content, importance, ttl_seconds, metadata)


class ProjectLayer(BaseMemoryLayer):
    """
    Project memory layer.

    Novel-level scope for story-wide information.
    Characters, world-building, plot structures, etc.
    No automatic expiration.
    """

    def __init__(
        self,
        project_id: Optional[str] = None,
        config: Optional[LayerConfig] = None
    ):
        super().__init__(config or LAYER_CONFIGS[LayerType.PROJECT])
        self.project_id = project_id

    async def store(
        self,
        id: str,
        content: str,
        importance: float = 0.5,
        ttl_seconds: Optional[int] = None,
        metadata: Optional[Dict[str, Any]] = None
    ) -> LayerEntry:
        metadata = metadata or {}
        if self.project_id:
            metadata["project_id"] = self.project_id

        return await super().store(id, content, importance, ttl_seconds, metadata)


class LayerManager:
    """
    Manager for all memory layers.

    Provides unified access to layer operations.
    """

    def __init__(
        self,
        session_id: Optional[str] = None,
        user_id: Optional[str] = None,
        project_id: Optional[str] = None
    ):
        self._layers: Dict[LayerType, BaseMemoryLayer] = {
            LayerType.EPHEMERAL: EphemeralLayer(),
            LayerType.SESSION: SessionLayer(session_id),
            LayerType.USER: UserLayer(user_id),
            LayerType.PROJECT: ProjectLayer(project_id),
        }
        logger.info("LayerManager initialized with all layers")

    def get_layer(self, layer_type: LayerType) -> BaseMemoryLayer:
        """Get a specific layer."""
        return self._layers[layer_type]

    async def store(
        self,
        layer_type: LayerType,
        id: str,
        content: str,
        importance: float = 0.5,
        ttl_seconds: Optional[int] = None,
        metadata: Optional[Dict[str, Any]] = None
    ) -> LayerEntry:
        """Store content in specified layer."""
        layer = self._layers[layer_type]
        return await layer.store(id, content, importance, ttl_seconds, metadata)

    async def retrieve(
        self,
        id: str,
        layer_type: Optional[LayerType] = None
    ) -> Optional[LayerEntry]:
        """
        Retrieve entry by ID.

        If layer_type is None, searches all layers.
        """
        if layer_type is not None:
            return await self._layers[layer_type].retrieve(id)

        # Search all layers (most persistent first)
        for lt in [LayerType.PROJECT, LayerType.USER, LayerType.SESSION, LayerType.EPHEMERAL]:
            entry = await self._layers[lt].retrieve(id)
            if entry is not None:
                return entry
        return None

    async def expire_all_layers(self) -> Dict[LayerType, int]:
        """Expire entries in all layers."""
        results = {}
        for lt, layer in self._layers.items():
            results[lt] = await layer.expire_all()
        return results

    async def stats_all(self) -> Dict[LayerType, LayerStats]:
        """Get stats for all layers."""
        results = {}
        for lt, layer in self._layers.items():
            results[lt] = await layer.stats()
        return results


def create_layer(layer_type: LayerType, **kwargs) -> BaseMemoryLayer:
    """
    Factory function to create a memory layer.

    Args:
        layer_type: Type of layer to create.
        **kwargs: Additional arguments for the layer.

    Returns:
        Configured memory layer instance.
    """
    layer_classes = {
        LayerType.EPHEMERAL: EphemeralLayer,
        LayerType.SESSION: SessionLayer,
        LayerType.USER: UserLayer,
        LayerType.PROJECT: ProjectLayer,
    }

    layer_class = layer_classes.get(layer_type)
    if layer_class is None:
        raise ValueError(f"Unknown layer type: {layer_type}")

    return layer_class(**kwargs)

File: src-ts/memory/memory-layers.ts
/**
 * Memory Layer Handlers - Four-Layer Memory Architecture
 *
 * Implements IMemoryLayer interface with 4 layer handlers:
 * - EphemeralLayer: Temporary memory (< 1 hour TTL)
 * - SessionLayer: Current task scope
 * - UserLayer: Long-term user preferences
 * - ProjectLayer: Novel-level scope
 *
 * Each layer has different TTL, scope, and persistence characteristics.
 */

/** Memory layer types with TTL characteristics. */
export enum LayerType {
  EPHEMERAL = "ephemeral", // < 1 hour
  SESSION = "session", // Current task
  USER = "user", // Long-term preferences
  PROJECT = "project", // Novel-level scope
}

/** Configuration for a memory layer. */
export class LayerConfig {
  layerType: LayerType;
  defaultTtlSeconds: number | null;
  maxEntries: number;
  autoExpire: boolean;
  persistOnClose: boolean;

  constructor(params: {
    layerType: LayerType;
    defaultTtlSeconds?: number | null;
    maxEntries?: number;
    autoExpire?: boolean;
    persistOnClose?: boolean;
  }) {
    this.layerType = params.layerType;
    this.defaultTtlSeconds = params.defaultTtlSeconds ?? null;
    this.maxEntries = params.maxEntries ?? 10000;
    this.autoExpire = params.autoExpire ?? true;
    this.persistOnClose = params.persistOnClose ?? true;
  }
}

/** Default configurations for each layer */
export const LAYER_CONFIGS: Record<LayerType, LayerConfig> = {
  [LayerType.EPHEMERAL]: new LayerConfig({
    layerType: LayerType.EPHEMERAL,
    defaultTtlSeconds: 3600, // 1 hour
    maxEntries: 1000,
    autoExpire: true,
    persistOnClose: false,
  }),
  [LayerType.SESSION]: new LayerConfig({
    layerType: LayerType.SESSION,
    defaultTtlSeconds: 86400, // 24 hours
    maxEntries: 5000,
    autoExpire: true,
    persistOnClose: true,
  }),
  [LayerType.USER]: new LayerConfig({
    layerType: LayerType.USER,
    defaultTtlSeconds: null, // No expiration
    maxEntries: 10000,
    autoExpire: false,
    persistOnClose: true,
  }),
  [LayerType.PROJECT]: new LayerConfig({
    layerType: LayerType.PROJECT,
    defaultTtlSeconds: null, // No expiration
    maxEntries: 50000,
    autoExpire: false,
    persistOnClose: true,
  }),
};

/** Entry stored in a memory layer. */
export class LayerEntry {
  id: string;
  content: string;
  createdAt: Date;
  expiresAt: Date | null;
  importance: number;
  accessCount: number;
  lastAccessed: Date | null;
  metadata: Record<string, unknown>;

  constructor(params: {
    id: string;
    content: string;
    createdAt?: Date;
    expiresAt?: Date | null;
    importance?: number;
    accessCount?: number;
    lastAccessed?: Date | null;
    metadata?: Record<string, unknown>;
  }) {
    this.id = params.id;
    this.content = params.content;
    this.createdAt = params.createdAt ?? new Date();
    this.expiresAt = params.expiresAt ?? null;
    this.importance = params.importance ?? 0.5;
    this.accessCount = params.accessCount ?? 0;
    this.lastAccessed = params.lastAccessed ?? null;
    this.metadata = params.metadata ?? {};
  }

  /** Check if entry has expired. */
  isExpired(): boolean {
    if (this.expiresAt === null) {
      return false;
    }
    return new Date() > this.expiresAt;
  }

  /** Update access tracking. */
  touch(): void {
    this.accessCount += 1;
    this.lastAccessed = new Date();
  }
}

/** Statistics for a memory layer. */
export interface LayerStats {
  layerType: LayerType;
  totalEntries: number;
  expiredEntries: number;
  totalAccessCount: number;
  avgImportance: number;
  oldestEntry: Date | null;
  newestEntry: Date | null;
}

/**
 * Interface for memory layer implementations.
 *
 * Each layer handles storage, retrieval, and expiration
 * of memories within its scope.
 */
export interface IMemoryLayer {
  /** Get the layer type. */
  readonly layerType: LayerType;

  /** Get layer configuration. */
  readonly config: LayerConfig;

  /**
   * Store content in this layer.
   *
   * @param id - Unique identifier for the entry.
   * @param content - Content to store.
   * @param importance - Importance score (0.0-1.0).
   * @param ttlSeconds - Time-to-live in seconds (overrides default).
   * @param metadata - Additional metadata.
   * @returns Created LayerEntry.
   */
  store(
    id: string,
    content: string,
    importance?: number,
    ttlSeconds?: number | null,
    metadata?: Record<string, unknown> | null
  ): Promise<LayerEntry>;

  /**
   * Retrieve an entry by ID.
   *
   * @param id - Entry identifier.
   * @returns LayerEntry or null if not found/expired.
   */
  retrieve(id: string): Promise<LayerEntry | null>;

  /**
   * Manually expire an entry.
   *
   * @param id - Entry identifier.
   * @returns True if expired, False if not found.
   */
  expire(id: string): Promise<boolean>;

  /**
   * Expire all entries that have passed their TTL.
   *
   * @returns Number of entries expired.
   */
  expireAll(): Promise<number>;

  /**
   * Get layer statistics.
   *
   * @returns LayerStats with current state.
   */
  stats(): Promise<LayerStats>;

  /**
   * List entries in this layer.
   *
   * @param limit - Maximum entries to return.
   * @param includeExpired - Whether to include expired entries.
   * @returns List of LayerEntry objects.
   */
  listEntries(limit?: number, includeExpired?: boolean): Promise<LayerEntry[]>;

  /**
   * Clear all entries from this layer.
   *
   * @returns Number of entries cleared.
   */
  clear(): Promise<number>;
}

/**
 * Base implementation for memory layers.
 *
 * Provides common functionality for all layer types.
 */
export class BaseMemoryLayer implements IMemoryLayer {
  protected _config: LayerConfig;
  protected _entries: Map<string, LayerEntry> = new Map();

  constructor(config: LayerConfig) {
    this._config = config;
    console.log(`Initialized ${config.layerType} layer`);
  }

  get layerType(): LayerType {
    return this._config.layerType;
  }

  get config(): LayerConfig {
    return this._config;
  }

  async store(
    id: string,
    content: string,
    importance: number = 0.5,
    ttlSeconds: number | null = null,
    metadata: Record<string, unknown> | null = null
  ): Promise<LayerEntry> {
    // Calculate expiration
    let expiresAt: Date | null = null;
    const ttl = ttlSeconds ?? this._config.defaultTtlSeconds;
    if (ttl !== null) {
      expiresAt = new Date(Date.now() + ttl * 1000);
    }

    const entry = new LayerEntry({
      id,
      content,
      importance,
      expiresAt,
      metadata: metadata ?? {},
    });

    // Check capacity
    if (this._entries.size >= this._config.maxEntries) {
      await this._evictOldest();
    }

    this._entries.set(id, entry);
    return entry;
  }

  async retrieve(id: string): Promise<LayerEntry | null> {
    const entry = this._entries.get(id);
    if (entry === undefined) {
      return null;
    }

    if (entry.isExpired()) {
      if (this._config.autoExpire) {
        this._entries.delete(id);
      }
      return null;
    }

    entry.touch();
    return entry;
  }

  async expire(id: string): Promise<boolean> {
    if (this._entries.has(id)) {
      this._entries.delete(id);
      return true;
    }
    return false;
  }

  async expireAll(): Promise<number> {
    const expiredIds: string[] = [];
    for (const [id, entry] of this._entries) {
      if (entry.isExpired()) {
        expiredIds.push(id);
      }
    }
    for (const id of expiredIds) {
      this._entries.delete(id);
    }

    if (expiredIds.length > 0) {
      console.log(
        `Expired ${expiredIds.length} entries from ${this.layerType} layer`
      );
    }
    return expiredIds.length;
  }

  async stats(): Promise<LayerStats> {
    const entries = Array.from(this._entries.values());
    const expiredCount = entries.filter((e) => e.isExpired()).length;
    const totalAccess = entries.reduce((sum, e) => sum + e.accessCount, 0);
    const avgImportance =
      entries.length > 0
        ? entries.reduce((sum, e) => sum + e.importance, 0) / entries.length
        : 0.0;

    const createdTimes = entries.map((e) => e.createdAt);
    const oldest = createdTimes.length > 0 ? new Date(Math.min(...createdTimes.map((d) => d.getTime()))) : null;
    const newest = createdTimes.length > 0 ? new Date(Math.max(...createdTimes.map((d) => d.getTime()))) : null;

    return {
      layerType: this.layerType,
      totalEntries: entries.length,
      expiredEntries: expiredCount,
      totalAccessCount: totalAccess,
      avgImportance,
      oldestEntry: oldest,
      newestEntry: newest,
    };
  }

  async listEntries(
    limit: number = 100,
    includeExpired: boolean = false
  ): Promise<LayerEntry[]> {
    let entries = Array.from(this._entries.values());

    if (!includeExpired) {
      entries = entries.filter((e) => !e.isExpired());
    }

    // Sort by importance (descending), then by createdAt (descending)
    entries.sort((a, b) => {
      if (a.importance !== b.importance) {
        return b.importance - a.importance;
      }
      return b.createdAt.getTime() - a.createdAt.getTime();
    });
    return entries.slice(0, limit);
  }

  async clear(): Promise<number> {
    const count = this._entries.size;
    this._entries.clear();
    console.log(`Cleared ${count} entries from ${this.layerType} layer`);
    return count;
  }

  /** Evict oldest entry to make room. */
  protected async _evictOldest(): Promise<void> {
    if (this._entries.size === 0) {
      return;
    }

    let oldestId = "";
    let oldestTime = Infinity;

    for (const [id, entry] of this._entries) {
      if (entry.createdAt.getTime() < oldestTime) {
        oldestTime = entry.createdAt.getTime();
        oldestId = id;
      }
    }

    if (oldestId) {
      this._entries.delete(oldestId);
    }
  }
}

/**
 * Ephemeral memory layer.
 *
 * Short-lived memories with < 1 hour TTL.
 * Used for temporary context, scratch data, and intermediate results.
 */
export class EphemeralLayer extends BaseMemoryLayer {
  constructor(config?: LayerConfig) {
    super(config ?? LAYER_CONFIGS[LayerType.EPHEMERAL]);
  }

  async store(
    id: string,
    content: string,
    importance: number = 0.5,
    ttlSeconds: number | null = null,
    metadata: Record<string, unknown> | null = null
  ): Promise<LayerEntry> {
    // Ephemeral layer enforces max TTL of 1 hour
    const maxTtl = 3600;
    if (ttlSeconds !== null) {
      ttlSeconds = Math.min(ttlSeconds, maxTtl);
    } else {
      ttlSeconds = this._config.defaultTtlSeconds ?? maxTtl;
    }

    return super.store(id, content, importance, ttlSeconds, metadata);
  }
}

/**
 * Session memory layer.
 *
 * Task-scoped memories that persist for the current session.
 * Used for conversation context, task state, and working memory.
 */
export class SessionLayer extends BaseMemoryLayer {
  sessionId: string | null;

  constructor(sessionId?: string | null, config?: LayerConfig) {
    super(config ?? LAYER_CONFIGS[LayerType.SESSION]);
    this.sessionId = sessionId ?? null;
  }

  async store(
    id: string,
    content: string,
    importance: number = 0.5,
    ttlSeconds: number | null = null,
    metadata: Record<string, unknown> | null = null
  ): Promise<LayerEntry> {
    // Add sessionId to metadata
    const meta = metadata ?? {};
    if (this.sessionId) {
      meta["session_id"] = this.sessionId;
    }

    return super.store(id, content, importance, ttlSeconds, meta);
  }
}

/**
 * User memory layer.
 *
 * Long-term user preferences and learned behaviors.
 * No automatic expiration, persists across sessions.
 */
export class UserLayer extends BaseMemoryLayer {
  userId: string | null;

  constructor(userId?: string | null, config?: LayerConfig) {
    super(config ?? LAYER_CONFIGS[LayerType.USER]);
    this.userId = userId ?? null;
  }

  async store(
    id: string,
    content: string,
    importance: number = 0.5,
    ttlSeconds: number | null = null,
    metadata: Record<string, unknown> | null = null
  ): Promise<LayerEntry> {
    // User layer typically doesn't expire
    if (ttlSeconds === null) {
      ttlSeconds = this._config.defaultTtlSeconds; // null
    }

    const meta = metadata ?? {};
    if (this.userId) {
      meta["user_id"] = this.userId;
    }

    return super.store(id, content, importance, ttlSeconds, meta);
  }
}

/**
 * Project memory layer.
 *
 * Novel-level scope for story-wide information.
 * Characters, world-building, plot structures, etc.
 * No automatic expiration.
 */
export class ProjectLayer extends BaseMemoryLayer {
  projectId: string | null;

  constructor(projectId?: string | null, config?: LayerConfig) {
    super(config ?? LAYER_CONFIGS[LayerType.PROJECT]);
    this.projectId = projectId ?? null;
  }

  async store(
    id: string,
    content: string,
    importance: number = 0.5,
    ttlSeconds: number | null = null,
    metadata: Record<string, unknown> | null = null
  ): Promise<LayerEntry> {
    const meta = metadata ?? {};
    if (this.projectId) {
      meta["project_id"] = this.projectId;
    }

    return super.store(id, content, importance, ttlSeconds, meta);
  }
}

/**
 * Manager for all memory layers.
 *
 * Provides unified access to layer operations.
 */
export class LayerManager {
  private _layers: Map<LayerType, BaseMemoryLayer>;

  constructor(
    sessionId?: string | null,
    userId?: string | null,
    projectId?: string | null
  ) {
    this._layers = new Map([
      [LayerType.EPHEMERAL, new EphemeralLayer()],
      [LayerType.SESSION, new SessionLayer(sessionId ?? null)],
      [LayerType.USER, new UserLayer(userId ?? null)],
      [LayerType.PROJECT, new ProjectLayer(projectId ?? null)],
    ]);
    console.log("LayerManager initialized with all layers");
  }

  /** Get a specific layer. */
  getLayer(layerType: LayerType): BaseMemoryLayer {
    const layer = this._layers.get(layerType);
    if (!layer) {
      throw new Error(`Unknown layer type: ${layerType}`);
    }
    return layer;
  }

  /** Store content in specified layer. */
  async store(
    layerType: LayerType,
    id: string,
    content: string,
    importance: number = 0.5,
    ttlSeconds?: number | null,
    metadata?: Record<string, unknown> | null
  ): Promise<LayerEntry> {
    const layer = this._layers.get(layerType);
    if (!layer) {
      throw new Error(`Unknown layer type: ${layerType}`);
    }
    return layer.store(id, content, importance, ttlSeconds, metadata);
  }

  /**
   * Retrieve entry by ID.
   *
   * If layerType is null, searches all layers.
   */
  async retrieve(
    id: string,
    layerType?: LayerType | null
  ): Promise<LayerEntry | null> {
    if (layerType != null) {
      const layer = this._layers.get(layerType);
      return layer ? layer.retrieve(id) : null;
    }

    // Search all layers (most persistent first)
    const searchOrder = [
      LayerType.PROJECT,
      LayerType.USER,
      LayerType.SESSION,
      LayerType.EPHEMERAL,
    ];
    for (const lt of searchOrder) {
      const layer = this._layers.get(lt);
      if (layer) {
        const entry = await layer.retrieve(id);
        if (entry !== null) {
          return entry;
        }
      }
    }
    return null;
  }

  /** Expire entries in all layers. */
  async expireAllLayers(): Promise<Record<LayerType, number>> {
    const results: Partial<Record<LayerType, number>> = {};
    for (const [lt, layer] of this._layers) {
      results[lt] = await layer.expireAll();
    }
    return results as Record<LayerType, number>;
  }

  /** Get stats for all layers. */
  async statsAll(): Promise<Record<LayerType, LayerStats>> {
    const results: Partial<Record<LayerType, LayerStats>> = {};
    for (const [lt, layer] of this._layers) {
      results[lt] = await layer.stats();
    }
    return results as Record<LayerType, LayerStats>;
  }
}

/** Layer constructor parameters */
type LayerConstructorParams = {
  sessionId?: string;
  userId?: string;
  projectId?: string;
  config?: LayerConfig;
};

/**
 * Factory function to create a memory layer.
 *
 * @param layerType - Type of layer to create.
 * @param kwargs - Additional arguments for the layer.
 * @returns Configured memory layer instance.
 */
export function createLayer(
  layerType: LayerType,
  kwargs: LayerConstructorParams = {}
): BaseMemoryLayer {
  switch (layerType) {
    case LayerType.EPHEMERAL:
      return new EphemeralLayer(kwargs.config);
    case LayerType.SESSION:
      return new SessionLayer(kwargs.sessionId, kwargs.config);
    case LayerType.USER:
      return new UserLayer(kwargs.userId, kwargs.config);
    case LayerType.PROJECT:
      return new ProjectLayer(kwargs.projectId, kwargs.config);
    default:
      throw new Error(`Unknown layer type: ${layerType}`);
  }
}
